src/cmflinel.cpp

```diff
@@ -0,0 +1,525 @@
+/*
+ * Linel CMF Player --  Copyright (c) 2017 Sebastian Kienzl <seb@knzl.de>
+ *
+ * Part of Adplug - Replayer for many OPL2/OPL3 audio file formats.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * 
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+/*
+ * Most songs in this format use rhythm mode.
+ * The emus don't support rhythm mode too well, "nuked" seems good.
+ * I didn't verify with real hardware, so I don't know how it's supposed to sound.
+ */
+
+#include <string.h>
+#include <stddef.h>
+#include "cmflinel.h"
+#include "debug.h"
+
+CPlayer* CcmfLinelPlayer::factory(Copl* newopl)
+{
+	return new CcmfLinelPlayer(newopl);
+}
+
+static const int16_t fNumbers[] = { 0x157, 0x16b, 0x181, 0x198, 0x1b0, 0x1ca, 0x1e5, 0x202, 0x221, 0x242, 0x264, 0x288 };
+
+// Register offset for slot, e.g. 0x20 + slotRegisterOffsets[slotNr]
+static const int8_t slotRegisterOffsets[] = {
+	0,   1,  2,  3,  4,  5,
+	8,   9, 10, 11, 12, 13,
+	16, 17, 18, 19, 20, 21
+};
+
+/*
+Channel     Slot
+          Op1  Op2
+===================
+ 0        0    3
+ 1        1    4
+ 2        2    5
+ 3        6    9
+ 4        7   10
+ 5        8   11
+ 6       12   15
+ 7       13   16
+ 8       14   17
+
+ Rhythm Mode
+-------------------
+ BD      12   15   <- like Channel 6 in "melody mode"
+ SD      16
+ TOM     14
+ TC      17
+ HH      13
+*/
+
+// Slots for Op1/Op2 for channel
+static const struct {
+	int8_t slotOp1;
+	int8_t slotOp2;
+}
+channelSlots[] = { {0, 3}, {1, 4}, {2, 5}, {6, 9}, {7, 10}, {8, 11}, {12, 15}, {13, 16}, {14, 17}};
+
+/*
+Note that
+ op_table[chan]     == slotRegisterOffets[channelSlots[chan].slotOp1]
+ op_table[chan] + 3 == slotRegisterOffets[channelSlots[chan].slotOp2]
+However, the slot-numbers are needed for rhythm mode, thus the extra tables.
+*/
+
+// Rhythm mode mapping: Which slots for which channel (note that the BD uses 12/15 like in melody mode)
+static const int8_t channelSlotsRhythm[] = { -1, -1, -1, -1, -1, -1, 12, 16, 14, 17, 13};
+
+enum {
+	CHANNEL_RHY_BD  = 6,
+	CHANNEL_RHY_SN  = 7,
+	CHANNEL_RHY_TOM = 8,
+	CHANNEL_RHY_TC  = 9,
+	CHANNEL_RHY_HH  = 10
+};
+
+CcmfLinelPlayer::CcmfLinelPlayer(Copl* newopl): CPlayer(newopl)
+{
+}
+
+std::string CcmfLinelPlayer::gettype() { return std::string("Linel CMF"); }
```

Comment:
Yeah it is long, but I'd like to use CcmfmacsoperaPlayer and cmfmacsopera.* anyway, because "Linel" doesn't appear anywhere else.
"cmfsoundfx" or "cmfsfx" is misleading and "cmfsfxmo" is just a pile of letters :)